switch/mbedtls/switch-mbedtls.patch

```diff
@@ -51,3 +51,104 @@ diff -NBaur mbedtls-2.14.1/library/net_sockets.c mbedtls-2.14.1.switch/library/n
  
      return( 0 );
  }
+diff -Naur mbedtls-2.16.6/library/timing.c mbedtls-2.16.6-patched/library/timing.c
+--- mbedtls-2.16.6/library/timing.c	2020-04-09 14:12:23.000000000 +0100
++++ mbedtls-2.16.6-patched/library/timing.c	2020-07-28 23:38:36.881476522 +0100
+@@ -40,7 +40,7 @@
+ 
+ #if !defined(unix) && !defined(__unix__) && !defined(__unix) && \
+     !defined(__APPLE__) && !defined(_WIN32) && !defined(__QNXNTO__) && \
+-    !defined(__HAIKU__)
++    !defined(__HAIKU__) && !defined(__SWITCH__)
+ #error "This module only works on Unix and Windows, see MBEDTLS_TIMING_C in config.h"
+ #endif
+ 
+@@ -73,6 +73,10 @@
+ 
+ #endif /* _WIN32 && !EFIX64 && !EFI32 */
+ 
++#ifdef __SWITCH__
++#include <switch.h>
++#endif
++
+ #if !defined(HAVE_HARDCLOCK) && defined(MBEDTLS_HAVE_ASM) &&  \
+     ( defined(_MSC_VER) && defined(_M_IX86) ) || defined(__WATCOMC__)
+ 
+@@ -216,6 +220,16 @@
+ }
+ #endif /* !HAVE_HARDCLOCK && _MSC_VER && !EFIX64 && !EFI32 */
+ 
++#if !defined(HAVE_HARDCLOCK) && defined(__SWITCH__)
++
++#define HAVE_HARDCLOCK
++unsigned long mbedtls_timing_hardclock( void )
++{
++    return armGetSystemTick();
++}
++
++#endif
++
+ #if !defined(HAVE_HARDCLOCK)
+ 
+ #define HAVE_HARDCLOCK
+@@ -313,6 +327,51 @@
+     }
+ }
+ 
++#ifdef __SWITCH__
++static bool timer_thread_created = false;
++static unsigned int alarmUs = 0;
++static Thread timerThread;
++static alignas(0x1000) char timerThreadStack[0x1000];
++
++static void timer_thread_func(void *param)
++{
++    (void) param;
++    usleep(alarmUs);
++    mbedtls_timing_alarmed = 1;
++} 
++
++static void __attribute__((destructor)) cleanup_timer_thread() 
++{
++    if(timer_thread_created)
++    {
++        threadWaitForExit(&timerThread);
++        threadClose(&timerThread);
++    }
++}
++
++void mbedtls_set_alarm( int seconds )
++{
++    if(seconds == 0)
++    {
++        mbedtls_timing_alarmed = 1;
++        return;
++    }
++
++    mbedtls_timing_alarmed = 0;
++    alarmUs = seconds * 1000000;
++
++    if(timer_thread_created)
++    {
++        threadWaitForExit(&timerThread);
```

Comment:
Is this the desired behavior? This logic is saying: if there's an ongoing timer, this is going to wait for *that* timer to finish first, then afterwards set the new one.

Reply:
I think the "correct" behaviour here is unspecified - on Windows it'll leave the current timer running _and_ start a new one(!) so whichever is first wins, and on Linux with alarm() it'll cancel the current timer first. I agree this is probably not what we want though, but I'm unsure how to implement it without just keeping a list of thread handles...really nobody should be using this API because it sucks anyway.